pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -365,3 +416,42 @@ func loadRegistryConf(configPath string) (*tomlConfig, error) {
 	err = toml.Unmarshal(configBytes, &config)
 	return config, err
 }
+
+// IsSecureIndex returns true if indexName matches an element of CIDRs. indexName can be of form `host:port` or `host`
+// where the `host` part can be either a domain name or an IP address. If it is a domain name, then it will be
+// resolved to IP addresses for matching. If resolution fails, false is returned.
+func (s *SysRegistries) IsSecureIndex(indexName string) bool {
+	if index, ok := s.HostPortConfigs[indexName]; ok {
+		return index.Secure
+	}
+
+	host, _, err := net.SplitHostPort(indexName)
+	if err != nil {
+		// assume indexName is of the form `host` without the port and go on.
+		host = indexName
+	}
+
+	addrs, err := net.LookupIP(host)
```

Comment:
To reiterate, this is a TOCTOU vulnerability: an attacker with network access can override DNS to point to a known- or expected-to-be “insecure CIDR range” (127.0.0.1) during this lookup of any host name, causing `IsSecureIndex` to return false — and subsequently the `docker://` transport will make a non-TLS connection with another DNS lookup, at which point the attacker can override DNS to point at the attacker’s server and the client can’t tell that it is not talking to the legitimate registry.

I’m very tempted to demand gating this behind a `sysregistriesv2.YesIKnowCIDRLookupsAreInsecureAndBreakTLSAndIAcceptFullResponsibility` (💀) variable that CRI-O would have to set to opt into using this.

Reply:
@mrunalp WDYT?